src/awst_build/eb/arc4/base.ts

```diff
@@ -85,3 +101,103 @@ class Arc4EqualsFunctionBuilder extends FunctionBuilder {
     )
   }
 }
+
+export class Arc4EncodedFromBytesFunctionBuilder extends FunctionBuilder {
+  constructor(
+    sourceLocation: SourceLocation,
+    private ptypeFactory?: (args: PType[]) => ARC4EncodedType,
+    private genericArgsCount?: number,
+  ) {
+    super(sourceLocation)
+  }
+  call(args: ReadonlyArray<NodeBuilder>, typeArgs: ReadonlyArray<PType>, sourceLocation: SourceLocation): NodeBuilder {
+    const funcName = 'fromBytes'
+    const {
+      ptypes,
+      args: [initialValueBuilder],
+    } = parseFunctionArgs({
+      args,
+      typeArgs,
+      genericTypeArgs: this.genericArgsCount ?? 1,
+      funcName,
+      argSpec: (a) => [a.required(bytesPType)],
+      callLocation: sourceLocation,
+    })
+
+    const ptype = (this.ptypeFactory ? this.ptypeFactory(ptypes) : ptypes[0]) as ARC4EncodedType
```

Comment:
we should use the `invariant` or `codeInvariant` function to check `ptypes[0]` is actually an ARC4EncodedType. The choice depends on whether the source of the ptype is something the user specified or not (code errors for user influenced stuff and internal if it's a compiler issue that shouldn't ever come up)

Reply:
It has been updated to use `codeInvariant` in this commit, 6c31fad.